modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpDownloader.java

```diff
@@ -0,0 +1,258 @@
+                state = ((GeoIpTaskState) r.getState()).copy();
+            }, e -> logger.error("could not remove task [" + GEOIP_DOWNLOADER + "]", e)));
+        }
+    }
+
+    void updateDatabases() throws IOException {
+        logger.info("updating geoip databases");
+        String data = httpClient.getString(endpoint + "?key=11111111-1111-1111-1111-111111111111");
+        List<Map<String, Object>> response = XContentHelper.convertToList(XContentType.JSON.xContent(), data, true);
+        for (Map<String, Object> res : response) {
+            processDatabase(res);
+        }
+    }
+
+    void processDatabase(Map<String, Object> databaseInfo) {
+        Map<String, Metadata> currentDatabases = state.getDatabases();
+        String name = databaseInfo.get("name").toString().replace(".gz", "");
+        boolean databaseInState = currentDatabases.containsKey(name);
+        String md5 = (String) databaseInfo.get("md5_hash");
+        if (databaseInState && Objects.equals(md5, currentDatabases.get(name).getMd5())) {
+            updateTimestamp(currentDatabases, name);
+            return;
+        }
+        logger.info("updating geoip database [" + name + "]");
+        String url = databaseInfo.get("url").toString();
+        try (InputStream is = httpClient.get(url)) {
+            int firstChunk = databaseInState ? currentDatabases.get(name).getLastChunk() + 1 : 0;
+            int lastChunk = indexChunks(name, is, firstChunk);
+            if (lastChunk > firstChunk) {
+                currentDatabases.put(name, new Metadata(System.currentTimeMillis(), firstChunk, lastChunk - 1, md5));
+                if (updateTaskState(name)) {
+                    logger.info("updated geoip database [" + name + "]");
+                }
+            }
+        } catch (Exception e) {
+            logger.error("error updating geoip database [" + name + "]", e);
+        }
+    }
+
+    //visible for testing
+    protected void updateTimestamp(Map<String, Metadata> currentDatabases, String name) {
+        Metadata old = currentDatabases.get(name);
+        currentDatabases.put(name, new Metadata(System.currentTimeMillis(), old.getFirstChunk(), old.getLastChunk(), old.getMd5()));
+        logger.info("geoip database [" + name + "] is up to date, updated timestamp");
+        updateTaskState(name);
+    }
+
+    boolean updateTaskState(String name) {
+        PlainActionFuture<PersistentTask<?>> future = PlainActionFuture.newFuture();
+        persistentTask.updatePersistentTaskState(state, future);
+        try {
+            state = ((GeoIpTaskState) future.actionGet().getState()).copy();
+            return true;
+        } catch (Exception e) {
+            logger.error("can not update state during geoip database [" + name + "] update", e);
+            return false;
+        }
+    }
+
+    //visible for testing
+    int indexChunks(String name, InputStream is, int chunk) throws IOException {
+        while (true) {
```

Comment:
maybe rewrite to a for loop: `for (byte[] buf = getChunk(is); buf.length != 0; buf = getChunk(is)) {`

Reply:
Clever! I've changed it